File: 206.反转链表.java
/*
 * @lc app=leetcode.cn id=206 lang=java
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        // 递归
        // if (head == null || head.next == null) {
        //     return head;
        // }
        // ListNode last = reverseList(head.next);
        // head.next.next = head;
        // head.next = null;
        // return last;

        if (head == null || head.next == null) {
            return head;
        }
        ListNode pre = head;
        ListNode tmp = null;
        head = head.next;
        pre.next = null;
        while (head != null) {
            tmp = head.next;
            head.next = pre;
            pre = head;
            head = tmp;
        }
        return pre;
    }
}
// @lc code=end

File: 206.反转链表.py
#
# @lc app=leetcode.cn id=206 lang=python3
#
# [206] 反转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # 递归
        # if not head or not head.next:
        #     return head
        # last = self.reverseList(head.next)
        # head.next.next = head
        # head.next = None
        # return last
        if not head or not head.next:
            return head
        pre = head
        head = head.next
        pre.next = None
        while head:
            tmp = head.next
            head.next = pre
            pre = head

            head = tmp

        return pre

# @lc code=end
